share/shwap/p2p/bitswap/block_fetch.go

```diff
@@ -0,0 +1,222 @@
+		if err := exchg.NotifyNewBlocks(ctx, bitswapBlk); err != nil {
+			log.Error("failed to notify new Bitswap blocks: %w", err)
+		}
+
+		blk, ok := duplicates[bitswapBlk.Cid()]
+		if !ok {
+			// common case: the block was populated by the hasher, so skip
+			continue
+		}
+		// uncommon duplicate case: concurrent fetching of the same block,
+		// so we have to unmarshal it ourselves instead of the hasher,
+		unmarshalFn := blk.UnmarshalFn(root)
+		_, err := unmarshal(unmarshalFn, bitswapBlk.RawData())
+		if err != nil {
+			// this means verification succeeded in the hasher but failed here
+			// this case should never happen in practice
+			// and if so something is really wrong
+			panic(fmt.Sprintf("unmarshaling duplicate block: %s", err))
+		}
+		// NOTE: This approach has a downside that we redo deserialization and computationally
+		// expensive computation for as many duplicates. We tried solutions that doesn't have this
+		// problem, but they are *much* more complex. Considering this a rare edge-case the tradeoff
+		// towards simplicity has been made.
+	}
+
+	return ctx.Err()
+}
+
+// unmarshal unmarshalls the Shwap Container data into a Block via UnmarshalFn
+// If unmarshalFn is nil -- gets it from the global unmarshalFns.
+func unmarshal(unmarshalFn UnmarshalFn, data []byte) ([]byte, error) {
+	var blk bitswappb.Block
+	err := blk.Unmarshal(data)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling block: %w", err)
+	}
+	cid, err := cid.Cast(blk.Cid)
+	if err != nil {
+		return nil, fmt.Errorf("casting cid: %w", err)
+	}
+	// getBlock ID out of CID validating it
+	id, err := extractCID(cid)
+	if err != nil {
+		return nil, fmt.Errorf("validating cid: %w", err)
+	}
+
+	if unmarshalFn == nil {
+		// getBlock registered UnmarshalFn and use it to check data validity and
+		// pass it to Fetch caller
+		val, ok := unmarshalFns.Load(cid)
+		if !ok {
+			return nil, fmt.Errorf("no unmarshallers registered for %s", cid.String())
+		}
+		entry := val.(*unmarshalEntry)
+
+		entry.Lock()
+		defer entry.Unlock()
+		unmarshalFn = entry.UnmarshalFn
+	}
+
+	err = unmarshalFn(blk.Container)
+	if err != nil {
+		return nil, fmt.Errorf("verifying data: %w", err)
+	}
+
+	return id, nil
+}
+
+// unmarshalFns exist to communicate between Fetch and hasher, and it's global as a necessity
+//
+// Fetch registers UnmarshalFNs that hasher then uses to validate and unmarshal Block responses coming
+// through Bitswap
+//
+// Bitswap does not provide *stateful* verification out of the box and by default
+// messages are verified by their respective MultiHashes that are registered globally.
+// For every Block type there is a global hasher registered that accesses stored UnmarshalFn once a
+// message is received. It then uses UnmarshalFn to validate and fill in the respective Block
+//
+// sync.Map is used to minimize contention for disjoint keys
+var unmarshalFns sync.Map
+
+type unmarshalEntry struct {
+	sync.Mutex
+	UnmarshalFn
+}
```

Comment:
It seems easy to forget why this needs to have mutex, lets add comments

Reply:
Agreed